### ISSUE TITLE
How to remove?

### ISSUE THREAD

**Issue by Anonymous941** [Zygo/bees#280]
I've run bees, and now the filesystem is reporting that there is no space, despite 1 gigabyte of free space.  I assume it's the metadata of bees, but I have no idea how to remove it.  How do I just remove the hash tables and other data that bees created?

**Comment by kakra** [comment]
You've probably run into an issue where bees first increases meta data usage on the filesystem before it can actually clean up and free space, and in that first phase, it ran into an issue with no more space left. You cannot remove this "meta data", it's not from bees, it's from the file system itself.

Try removing old snapshots if you do have some. It should clean up duplicated meta data, and may also free some partially shared extents. Then let bees continue so it can finally clean up and free actual data extents.

**Comment by lilydjwg** [comment]
> Try removing old snapshots if you do have some.

No. Don't do this if metadata has no space left. There is a bug that may permanently put the filesystem into an errored mode (read-only).

Try to run `btrfs balance start -dusage=10 mountpoint` to see if you can free up some block groups from the data part. If you can't, add a temporary disk and do that balance.

Once you have enough unallocated space (at least >1GiB; run `btrfs filesystem usage mountpoint` and look for `Device unallocated:`), you can start delete snapshots to release some metadata usage (run `btrfs filesystem usage mountpoint` and look for `Metadata,`). You can also delete files. You can balance some metadata if it's too sparse.

**Comment by kakra** [comment]
> No. Don't do this if metadata has no space left. There is a bug that may permanently put the filesystem into an errored mode (read-only).

Good point.

**Comment by Zygo** [comment]
The hash table is stored in `.beeshome` at the top level of the filesystem.  The default size is 1 GiB, so deleting that should help free up one block group.  Then follow the other advice--try balancing, adding space on another device, and removing _unshared_ files and snapshots--in that order.

**Comment by Anonymous941** [comment]
What ended up making my computer bootable again is running `btrfs filesystem resize 230g`, then `btrfs filesystem resize max`.  I then tried to run bees again, and the same problem occurred, despite me freeing about 40/250 gigabytes.  I'm not sure how to free more space, and don't want to delete all my snapshots

**Comment by lilydjwg** [comment]
You have 20GiB free space allocated as data block groups. Run something like `btrfs balance start -dusage=90 mountpoint` to get them out as unallocated so that it can be allocated as metadata.

You have pretty low free space anyway. Running bees on a lot of existing snapshots will occupy a significant amount of metadata space (and it takes a lot of time). I'm afraid that you may not have enough space to finish the initial run.

**Comment by Anonymous941** [comment]
It turns out that there are a lot of `btrbk` snapshots that I could delete.  But it still says that there is only `1.00 MiB` under `Device unallocated`

**Comment by Anonymous941** [comment]
Is it normal for `btrfs balance start` to take several minutes?  I thought it just started a background process

**Comment by Anonymous941** [comment]
After running, `Device unallocated` is now `21.00GiB`.  Start `bees` again

**Comment by Anonymous941** [comment]
It was able to complete its first run.  I'm leaving this issue open as I'm not sure if it's intended behavior to put the partition in this state

**Comment by bar-g** [comment]
Thank you for that explanation. A pretty involved situation.

In the end doesn't it still sound like a clear btrfs bug?: Not reserving (allocating) enough free space for metadata.

Isn't there at least a heuristic btrfs could use to cover (more) of those cases? Like whatever is larger: x GB per GB of device size plus something for each disk, or 16% of current metadata space.

**Comment by bronger** [comment]
But if I wait long enough before umount so that everything settled, deleting snapshots is a safe thing?

And then, where in e.g. beesstatus or in Bees log output can I see that Bees has finished its metadata clean-up in the sense that the temporary increase in metadata space consumption is over?

**Comment by kakra** [comment]
Maybe bees should have a check where it pauses if free space falls below a certain threshold (configurable with sane default), and then the admin could restart bees with an override switch if things look safe?

**Comment by Zygo** [comment]
> And then, where in e.g. beesstatus or in Bees log output can I see that Bees has finished its metadata clean-up in the sense that the temporary increase in metadata space consumption is over?

The snapshot metadata is permanently enlarged.  It goes away when the snapshots are removed.  New snapshots created after bees has run once don't have the expansion, because they're already deduplicated and don't need to be deduplicated again.

> Maybe bees should have a check where it pauses if free space falls below a certain threshold (configurable with sane default), and then the admin could restart bees with an override switch if things look safe?

I've had snapshot deletes fail with 90 GiB of free metadata space.  That's bigger than some users' entire filesystems.

I don't know if there's a workaround for this.  The real fix is breaking up snapshot deletes across multiple transactions in btrfs, but that requires some refactoring of the current snapshot delete code.

It may be possible to monitor metadata size and pause or warn if it increases too quickly; however, that's prone to false positives (what if I want to dedupe 20 snapshots, and have enough space for 20x metadata?), and it needs a notification mechanism to tell the operator that there's a problem.

With some fairly extensive metadata searching, it's possible to estimate how many metadata pages would be unshared by dedupe, and factor that into cost/gain calculations.  But in practice, it's much cheaper to just say "dedupe only extents newer than [transid]" where [transid] is the one where bees started running.

**Comment by bronger** [comment]
Thanks for the frank answer. IIRC, the difficulties on subvolume-extensive filesystems are already documented. So all one can really do is to keep an eye on free metadata space? It’s not elegant but doable.
